keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Dynamic Resource Allocation (DRA) was added to Kubernetes as an alpha feature in v1.26. It defines an alternative to the traditional device-plugin API for requesting access to third-party resources.
+
+
+By design, DRA uses parameters for resources that are completely
+opaque to core Kubernetes. They get interpreted by a DRA driver's controller
+(for allocating claims) and a DRA driver's kubelet plugin (for configuring
+resources a node). During scheduling of a pod, the kube-scheduler and any DRA
+driver controller(s) handling claims for the pod communicate back-and-forth through the
+apiserver by updating a `PodSchedulingContext` object, ultimately leading to the
+allocation of all pending claims and the pod being scheduled onto a node.
+
+This approach poses a problem for the [Cluster
+Autoscaler](https://github.com/kubernetes/autoscaler) (CA) or for any higher
+level controller that needs to make decisions for a group of pods (e.g. a job
+scheduler). It cannot simulate the effect of allocating or deallocating
+claims over time. Only the third-party DRA
+drivers have the information available to do this.
+
+"Semantic parameters" is an extension to DRA that addresses this problem by
+making claim parameters less opaque. Instead of managing the semantics of all
+claim parameters themselves, drivers now manage resources and describe them
+using a specific "semantic model" pre-defined by Kubernetes. This allows
+components aware of this "semantic model" to make decisions about these
+resources without outsourcing them to some third-party controller. For example,
+the scheduler is now able to allocate claims rapidly, without back-and-forth
+communication with DRA drivers.
+
+At a high-level, this extension takes the following form:
+
+* DRA drivers publish their available resources in the form of a
+  `NodeResourceSlice` object on a node-by-node basis according to one of the
+  builtin "semantic models" known to Kubernetes. This object is stored in the
+  API server and available to the scheduler (or Cluster Autoscaler) to query
+  when a resource request comes in later on.
+
+* When a user wants to consume a resource, they create a `ResourceClaim`,
+  which, in turn, references a claim parameters object. This object defines how
+  many resources are needed and which capabilities they must have. Typically, it
+  is defined using a vendor-specific type which might also support configuration
+  parameters (i.e. parameters that are *not* needed for allocation but *are*
+  needed for configuration).
+
+* With such a claim in place, DRA drivers "resolve" the contents of any
```

Comment:
> For consideration during impl: What if the resolved form of the vendor-specific params was in the status of the claim?

That was my original thought too. But Patrick pointed out that this would mean the vendor controllers all need write access to the claim status. We decided that the simplest way, from the point of view of RBAC, was the intermediate object, the `ResourceClaimParameters`. The vendor controllers can create and update those, which include the resolved claim parameters, plus the config data needed by the driver stored in a rawextension for pass through. The scheduler only needs read access to that resource, rather than to any vendor specific resources.

> Or another option, which I am not sure is better or worse UX: What if the Pod referenced a vendor-specific "claim resource" directly, and that claim resource's status referenced a generic claim? That makes the user's life slightly simpler, but relies on duck-typing.

Similar RBAC issues then, as now anything that needs to follow that reference needs RBAC to those vendor-specific objects.